mypy/semanal.py

```diff
@@ -655,23 +660,40 @@ def visit_class_def(self, defn: ClassDef) -> None:
         self.clean_up_bases_and_infer_type_variables(defn)
         if self.analyze_typeddict_classdef(defn):
             return
-        if self.analyze_namedtuple_classdef(defn):
-            # just analyze the class body so we catch type errors in default values
-            self.enter_class(defn)
+        named_tuple_info = self.analyze_namedtuple_classdef(defn)
+        if named_tuple_info is not None:
+            # temporarily clear the names dict so we don't get errors about duplicate names that
+            # were set in build_namedtuple_typeinfo
+            nt_names = named_tuple_info.names
+            named_tuple_info.names = SymbolTable()
+
+            self.bind_class_type_vars(named_tuple_info)
+            self.enter_class(named_tuple_info)
+
             defn.defs.accept(self)
+
             self.leave_class()
+            self.unbind_class_type_vars()
+
+            # make sure we didn't use illegal names, then reset the names in the typeinfo
+            for prohibited in NAMEDTUPLE_PROHIBITED_NAMES:
+                if prohibited in named_tuple_info.names:
+                    self.fail('Cannot overwrite NamedTuple attribute "{}"'.format(prohibited),
+                              named_tuple_info.names[prohibited].node)
+
+            named_tuple_info.names.update(nt_names)
```

Comment:
The sets will actually always overlap, because `build_namedtuple_typeinfo` inserts names for the namedtuple fields, and visiting the class body will insert those names again. I think that's safe because `build_namedtuple_typeinfo` is very restrictive in the kind of statements it accepts within a class body, but I'll add some more tests to verify that it's correct (you already suggested some below).